k8s/autopath: Add CNAMES

chrisohaver opened the pull request:
* I added unit tests for ServeDNS, and coverage for autopath queries that can be handled in the k8s middleware.  no coverage for queries that fall though to next middleware. #766 
* Autopath adds CNAME records to "bridge the gap" between header question and the A record.  #765
* Fixes the NXDOMAIN override to cause the client to stop traversing the search path when there is no result. (almost) #761
* The CNAME and NXDOMAIN items above are in part handled by a new ResponseWriter implementor: AutoPathWriter, which handles deferring writes to the outer middleware (k8s), conditionally rewriting the rcode, and restoring the header question. 

Work in Progress.  Some issues...
* disabling the NXDOMAIN feature doesn't work.  Its always enabled (not sure why yet).
* AAAA requests get A responses when passing them through the next middleware (proxy).
* currently only tries AutoPath for A and AAAA.  Need to test other types.
* Still no test coverage for queries that need to get answers from next middleware.

miekg commented on middleware/kubernetes/autopath.go:
```diff
@@ -0,0 +1,79 @@
+				Class:  dns.ClassINET,
+				Ttl:    a.Header().Ttl},
+			Target: dns.Fqdn(a.Header().Name)}
+		res.Answer = append(res.Answer, &cname)
```
Does appending, while ranging actually work

chrisohaver replied:
The way it behaves works for what we want to do... the records appended in the loop are not ranged over.

miekg commented on middleware/kubernetes/handler.go:
```diff
@@ -40,10 +38,8 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
-	if k.AutoPath.Enabled && k.IsNameError(err) {
+	if k.AutoPath.Enabled && k.IsNameError(err) && (state.QType() == dns.TypeA || state.QType() == dns.TypeAAAA) {
 		p := k.findPodWithIP(state.IP())
 		for p != nil {
```
`if p == nil { continue}`

chrisohaver replied:
Not sure I follow this comment.

chrisohaver replied:
if you mean to replace the `for p != nil {`  I don't think it would work. `continue` I thought only works inside a loop.

miekg commented on middleware/kubernetes/handler.go:
```diff
@@ -53,38 +49,46 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
 					break
 				}
 			}
 			if !k.IsNameError(err) {
```
Why is this IsNameErrror check repeated outside the `for`-loop as well?

chrisohaver replied:
This breaks out of the outer loop (`for p!= nil`).
I could label the outer loop, and break out from the inner loop line 60 if you prefer.

miekg commented on middleware/kubernetes/handler.go:
```diff
@@ -53,38 +49,46 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
-			// Search "svc.cluster.local" and "cluster.local"
+			origQName := state.QName()
+			// Search "svc.cluster.local." and "cluster.local."
 			for i := 0; i < 2; i++ {
```
Why is this looping twice over the same thing? Unless my eyes are deceiving me?

chrisohaver replied:
it strips the first segment from path each time through.
The original path is ... `w.x.y.z`, then the first loop does `x.y.z`, and the second does `y.z`

miekg replied:
On what line in that loop is `i` used?

chrisohaver replied:
`i` is only used in the loop condition.

miekg replied:
I still don't see it :-( Where is the path changed for the 2nd run?

chrisohaver replied:
first line in the loop does, effectively `path = path minus first segment` ... 
```
path = strings.Join(dns.SplitDomainName(path)[1:], ".")
```

miekg replied:
ah, sorry. yes.

chrisohaver commented:
The range is established once, before the loop starts. Appended items don't get ranged over, since they are outside the initial range.

Sent from my mobile

On Jul 5, 2017, at 8:46 AM, Miek Gieben <notifications@github.com<mailto:notifications@github.com>> wrote:


@miekg commented on this pull request.

________________________________

In middleware/kubernetes/autopath.go<https://github.com/coredns/coredns/pull/771#discussion_r125632310>:

> +func (r *AutoPathWriter) WriteMsg(res *dns.Msg) error {
+       if res.Rcode == dns.RcodeNameError {
+               res.Rcode = r.Rcode
+       }
+       for _, a := range res.Answer {
+               if r.original.Name == a.Header().Name {
+                       continue
+               }
+               cname := dns.CNAME{
+                       Hdr: dns.RR_Header{
+                               Name:   r.original.Name,
+                               Rrtype: dns.TypeCNAME,
+                               Class:  dns.ClassINET,
+                               Ttl:    a.Header().Ttl},
+                       Target: dns.Fqdn(a.Header().Name)}
+               res.Answer = append(res.Answer, &cname)


Does appending, while ranging actually work

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/coredns/coredns/pull/771#pullrequestreview-48039372>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ANq-NVfVRM69SlluAEXfWzOxFPC4FP_fks5sK4WIgaJpZM4OLd2X>.

miekg commented on middleware/kubernetes/handler.go:
```diff
@@ -53,38 +49,46 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
-				r = state.NewWithQuestion(strings.Join([]string{name, hostsearch}, "."), state.QType()).Req
-				rcode, nextErr := middleware.NextOrFailure(k.Name(), k.Next, ctx, wr, r)
-				if rcode == dns.RcodeSuccess {
+				state = state.NewWithQuestion(strings.Join([]string{name, hostsearch}, "."), state.QType())
```
We might want to have a dnsutil.Join function for these?

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: k8s/auto..." ]
>>  			for i := 0; i < 2; i++ {
> 				path = strings.Join(dns.SplitDomainName(path)[1:], ".")
> 				state = state.NewWithQuestion(strings.Join([]string{name, path}, "."), state.QType())
> 				records, extra, _, err = k.routeRequest(zone, state)
> 				if !k.IsNameError(err) {
>+					records = append(records, NewCNAME(origQName, records[0].Header().Name, records[0].Header().Ttl))
> 					break
> 				}
> 			}
> 			if !k.IsNameError(err) {
>
>This breaks out of the outer loop (`for p!= nil`).
>I could label the outer loop, and break out from the inner loop line 60 if you prefer.

OK, just leave it like this then.


/Miek

--
Miek Gieben

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] WIP: k8s/auto..." ]
>chrisohaver commented on this pull request.
>
>
>
>>  	records, extra, _, err := k.routeRequest(zone, state)
>-
>-	if k.AutoPath.Enabled && k.IsNameError(err) {
>+	if k.AutoPath.Enabled && k.IsNameError(err) && (state.QType() == dns.TypeA || state.QType() == dns.TypeAAAA) {
> 		p := k.findPodWithIP(state.IP())
> 		for p != nil {
>
>if you mean to replace the `for p != nil {`  I don't think it would work. `continue` I thought only works inside a loop.

True... I was just wondering if there was an "early out" here, because all the 
work happens in the if-body. Maybe not.

chrisohaver commented:
I've fixed the remaining bigger issues...
* OnNXDOMAIN fixed, unit test added.
* AAAA/A response switch fixed, unit test added.
* Added coverage for next middleware cases .

Only A and AAAA can take advantage of the feature for now.  I think we should leave that as is, and add other types later.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] k8s/autopath:..." ]
>I've fixed the remaining bigger issues...
>* OnNXDOMAIN fixed, unit test added.
>* AAAA/A response switch fixed, unit test added.
>* Added coverage for next middleware cases .
>
>Only A and AAAA can take advantage of the feature for now.  I think we should leave that as is, and add other types later.

Seems legit, that's probably 90% of the queries anyway.

miekg commented on middleware/kubernetes/README.md:
```diff
@@ -168,6 +168,7 @@ specified).
     ;google.com.default.svc.cluster.local. IN A
 
     ;; ANSWER SECTION:
+    google.com.default.svc.cluster.local. 211 IN CNAME google.com.
```
Where does the TTL come from?

chrisohaver replied:
It's copied from the A record.

miekg replied:
Why is the TTL on the A record diff. then?

chrisohaver replied:
Ah - good eye... I'll fix the README.

miekg commented on middleware/kubernetes/autopath.go:
```diff
@@ -0,0 +1,82 @@
+// OverrideMsg records the status code and calls the
+// underlying ResponseWriter's WriteMsg method unless
+// the write is deferred, or force = true.
+func (apw *AutoPathWriter) OverrideMsg(res *dns.Msg, force bool) error {
```
It is probably cleaner to have two separately named functions that do the 2 things the bool differentiates between. Like you have before

chrisohaver replied:
There is WriteMsg, and ForceWriteMsg.  The first calls OverrideMsg with force=false. the second calls OverrideMsg with force=true.

I should unexpose OverrideMsg, might make is clearer.

miekg commented on middleware/kubernetes/kubernetes.go:
```diff
@@ -57,6 +57,10 @@ type AutoPath struct {
 	OnNXDOMAIN     int
 }
 
+type Proxyer interface {
```
Why the new interface?

chrisohaver replied:
ah - I was starting to mock out the upstream lookups so those cases could be tested.

I can remove it - its not directly related to this feature.

miekg commented on middleware/kubernetes/handler.go:
```diff
@@ -115,6 +121,10 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
 	return dns.RcodeSuccess, nil
 }
 
+func NewCNAME(name string, target string, ttl uint32) *dns.CNAME {
```
You could use this in `OverrideMsg` as well. But this function duplicates some of the things that already exists in `etcd/msg/service.go` where `NewXXX` is defined for a bunch a RR types. Sad thing is they are methods on *Service. Probably needs to be factor out and put into `dnsutil`. Can you add a  TODO here that this needs to be done?

Also the function does not need to be expored

miekg commented on middleware/test/helpers.go:
```diff
@@ -279,6 +279,25 @@ func NextHandler(rcode int, err error) Handler {
 
+// MockHandler returns a Handler that returns an answer for the quesion in the
+// request per the question->answer map qMap.
+func MockHandler(qMap map[dns.Question]dns.Msg) Handler {
```
Please move this to `kubernetes/handler_test.go` and call it "autoPathHandler" or "testHandler"?

chrisohaver replied:
I considered it to have a general purpose - not specific to autopath or kubernetes.

miekg commented on request/request.go:
```diff
@@ -29,7 +29,7 @@ type Request struct {
 func (r *Request) NewWithQuestion(name string, typ uint16) Request {
 	req1 := Request{W: r.W, Req: r.Req.Copy()}
-	req1.Req.Question[0] = dns.Question{Name: dns.Fqdn(name), Qtype: dns.ClassINET, Qclass: typ}
+	req1.Req.Question[0] = dns.Question{Name: dns.Fqdn(name), Qclass: dns.ClassINET, Qtype: typ}
```
Ouch

chrisohaver replied:
This was the cause of the A/AAAA swap.
Class `IN` uses the same numerical value as type `A`. So when an `AAAA` record came in, this function was turning it into a `A` record with an (invalid) class `28` ... clients seem to ignore the invalid class.

chrisohaver commented on middleware/kubernetes/autopath.go:
```diff
@@ -0,0 +1,82 @@
+		res.Answer = append(res.Answer, &cname)
+	}
+	res.Question[0] = apw.original
+	if res.Rcode != dns.RcodeSuccess && !force {
```
I should move this check before the for loop.

miekg commented:
Ah ok. True looks useful. Better name than Mock would help then :)

On Jul 7, 2017 2:46 PM, "chrisohaver" <notifications@github.com> wrote:

*@chrisohaver* commented on this pull request.
------------------------------

In middleware/test/helpers.go
<https://github.com/coredns/coredns/pull/771#discussion_r126150071>:

> @@ -279,6 +279,25 @@ func NextHandler(rcode int, err error) Handler {
 	})
 }

+// MockHandler returns a Handler that returns an answer for the quesion in the
+// request per the question->answer map qMap.
+func MockHandler(qMap map[dns.Question]dns.Msg) Handler {

I considered it to have a general purpose - not specific to autopath or
kubernetes.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/coredns/coredns/pull/771#discussion_r126150071>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAVkW1ALX7vXgM24kuOY5fZjoCvQ4Ss8ks5sLjahgaJpZM4OLd2X>
.

johnbelamaric commented:
@miekg you want to hold off on the merge? I was thinking this was holding 009.

miekg commented:
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] k8s/autopath:..." ]
>@miekg you want to hold off on the merge? I was thinking this was holding 009.

Yes please. If the middleware *autopath* actually works, we might want to revert 
a few commits even...

miekg commented on middleware/kubernetes/handler.go:
```diff
@@ -53,38 +51,46 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
+			newstate := state.NewWithQuestion(strings.Join([]string{name, "."}, ""), state.QType())
+			records, extra, _, err = k.routeRequest(zone, newstate)
 			if !k.IsNameError(err) {
+				records = append(records, newCNAME(origQName, records[0].Header().Name, records[0].Header().Ttl))
```
you need a put the CNAME first and then the original A record otherwise glibc won't understand it.

johnbelamaric commented on middleware/kubernetes/autopath.go:
```diff
@@ -55,7 +55,9 @@ func (apw *AutoPathWriter) overrideMsg(res *dns.Msg, force bool) error {
 			continue
 		}
-		res.Answer = append(res.Answer, newCNAME(apw.original.Name, dns.Fqdn(a.Header().Name), a.Header().Ttl))
+		res.Answer = append(res.Answer, nil)
```
Maybe this instead: https://codingair.wordpress.com/2014/07/18/go-appendprepend-item-into-slice/

?

chrisohaver replied:
I actually googled up the same link.  But after reading comment ...

```
A better prepend, as it generates less garbage:

data = append(data, “”)
copy(data[1:], data)
data[0] = “Prepend Item”
```

I opted for the less garbagey method.  I did not verify if it produces less garbage, but it looks like it does.
